pkg/prm/docker.go

```diff
@@ -1,48 +1,318 @@
 	// TODO
 	return FAILURE, nil
 }
 
-func (*Docker) Exec(tool *Tool, args []string) (ToolExitCode, error) {
-	// TODO
-	return FAILURE, nil
+func (d *Docker) Exec(tool *Tool, args []string, prmConfig Config, paths DirectoryPaths) (ToolExitCode, error) {
+	// is Docker up and running?
+	status := d.Status()
+	if !status.IsAvailable {
+		log.Error().Msgf("Docker is not available")
+		return FAILURE, fmt.Errorf("%s", status.StatusMsg)
+	}
+
+	// clean up paths
+	codeDir, _ := filepath.Abs(paths.codeDir)
+	log.Info().Msgf("Code path: %s", codeDir)
+	cacheDir, _ := filepath.Abs(paths.cacheDir)
+	log.Info().Msgf("Cache path: %s", cacheDir)
+
+	log.Info().Msgf("Additional Args: %v", args)
+
+	// stand up a container
+	containerConf := container.Config{
+		Image: d.ImageName(tool, prmConfig),
+		Tty:   false,
+	}
+	// args can override the default CMD
+	if len(args) > 0 {
+		containerConf.Cmd = args
+	}
+
+	resp, err := d.OrigClient.ContainerCreate(d.Context,
+		&containerConf,
+		&container.HostConfig{
+			Mounts: []mount.Mount{
+				{
+					Type:   mount.TypeBind,
+					Source: codeDir,
+					Target: "/code",
+				},
+				{
+					Type:   mount.TypeBind,
+					Source: cacheDir,
+					Target: "/cache",
+				},
+			},
+		}, nil, nil, "")
+
+	if err != nil {
+		return FAILURE, err
+	}
+	// the autoremove functionality is too aggressive
+	// it fires before we can get at the logs
+	defer func() {
+		err := d.OrigClient.ContainerRemove(d.Context, resp.ID, types.ContainerRemoveOptions{
+			RemoveVolumes: true,
+		})
+		if err != nil {
+			log.Error().Msgf("Error removing container: %s", err)
+		}
+	}()
+
+	if err := d.OrigClient.ContainerStart(d.Context, resp.ID, types.ContainerStartOptions{}); err != nil {
+		return FAILURE, err
+	}
+
+	isDone := make(chan bool)
+	// Move this wait into a goroutine
+	// when its finished it will return and post to the isDone channel
+	go func(d *Docker, isDone chan bool) {
+		statusCh, errCh := d.OrigClient.ContainerWait(d.Context, resp.ID, container.WaitConditionNotRunning)
+		select {
+		case <-errCh:
+			isDone <- true
+		case <-statusCh:
+			isDone <- true
+		}
+	}(d, isDone)
+
+	// parse out the containers logs while we wait for the container to finish
+	for {
+		out, err := d.OrigClient.ContainerLogs(d.Context, resp.ID, types.ContainerLogsOptions{ShowStdout: true, ShowStderr: true, Tail: "2", Follow: true})
+		if err != nil {
+			return FAILURE, err
+		}
+
+		_, err = stdcopy.StdCopy(os.Stdout, os.Stderr, out)
+		if err != nil {
+			return FAILURE, err
+		}
+
+		if done := <-isDone; done {
+			return SUCCESS, nil
+		}
+	}
 }
 
-// nolint:unused
 func (d *Docker) initClient() (err error) {
 	if d.Client == nil {
 		cli, err := dockerClient.NewClientWithOpts(dockerClient.FromEnv)
 
+		if err != nil {
+			return err
+		}
+
 		d.Client = cli
-		return err
+		d.OrigClient = cli // TODO: remove this when we know all the functions that need added to the interface
+		d.Context = context.Background()
```

Comment:
We could potentially create a context with a timeout, like this:

```go
		ctx, _ := context.WithTimeout(context.Background(), 30*time.Minute)
		d.Context = ctx
```

Which would mean, if for some reason the script / app running in the tool were to hang, we'll handle that scenario.

Having said that, running all the spec tests on the `puppetlabs-apache` module single threaded is...quite a lot of time.

Reply:
This highlights that we need a `--timeout` flag or similar on exec, I think, but we can implement the 30m timeout and file an issue to implement the flag imo.

Reply:
please raise a ticket to add a timeout flag